File: 编译原理与设计/实验/Lab1.程序设计语言认知实验/卜梦煜-1120192419/Java/src/MatrixMultiply.java
import java.util.Scanner;
import java.io.*;

public class MatrixMultiply{
    public static void main(String[] args) throws IOException {

        // 计时
        long start = System.currentTimeMillis();

        // 输入模块
        File file_in = new File("data.txt");
        Scanner scanner = new Scanner(file_in);
        int n1 = scanner.nextInt();
        int[][] matrix1 = new int[n1][n1];
        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n1; j++) {
                matrix1[i][j] = scanner.nextInt();
            }
        }
        int n2 = scanner.nextInt();
        int[][] matrix2 = new int[n2][n2];
        for (int i = 0; i < n2; i++) {
            for (int j = 0; j < n2; j++) {
                matrix2[i][j] = scanner.nextInt();
            }
        }
        scanner.close();

        // 计算模块
        int [][] matrix3 = new int[n1][n2];
        for(int i = 0; i < n1; i++){
            for(int j = 0; j < n2; j++){
                matrix3[i][j] = 0;
                for(int k = 0; k < n1; k++){
                    matrix3[i][j] += matrix1[i][k] * matrix2[k][j];
                }
            }
        }

        // 输出模块
        File file_out = new File("result_java.txt");
        OutputStreamWriter osw = new OutputStreamWriter(new FileOutputStream(file_out));
        BufferedWriter bw = new BufferedWriter(osw);
        for (int i = 0; i < n1; i++) {
            StringBuffer sb = new StringBuffer();
            for (int j = 0; j < n2; j++) {
                sb.append(matrix3[i][j] + (j == n2 - 1 ? "\n" : " "));
            }
            bw.write(sb.toString());
        }
        bw.close();

        // 计时
        long end = System.currentTimeMillis();
        System.out.printf("Total time:%dms", end - start);
        return;
    }
}

File: 编译原理与设计/实验/Lab1.程序设计语言认知实验/卜梦煜-1120192419/Haskell/src/MatrixMultiply.hs
import Data.List (transpose)
import Data.Time.Clock

matrixMultiply :: [[Int]] -> [[Int]] -> [[Int]]
-- $为函数应用符，用于降低函数优先级，具有右结合性
-- transpose：矩阵转置
matrixMultiply a b = [ [ sum $ zipWith (*) x y | y <- transpose b] | x <- a ]

readMatrix :: [String] -> [[Int]]
readMatrix a = f a
    where
        f = map $ map read . words


main = do
    -- 计时
    start <- getCurrentTime

    -- 输入模块
    let file_in = "data.txt"
    content <- readFile file_in
    let line = lines content
        n = read (head line) :: Int
        a = tail $ take (n + 1) line
        b = drop (n + 2) line

    -- 运算模块
    let matrix1 = readMatrix a
        matrix2 = readMatrix b
        matrix3 = matrixMultiply matrix1 matrix2

    -- 输出模块
    -- unwords函数将输入的列表连接成字符串
    let temp = map (unwords . map show) matrix3
        out = unlines temp
        file_out = "result_haskell.txt"
    writeFile file_out out

    --计时
    end <- getCurrentTime
    print $ diffUTCTime end start
